File: 백준/Gold/1655. 가운데를 말해요/가운데를 말해요.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        PriorityQueue<Integer> left = new PriorityQueue<>(Collections.reverseOrder());
        PriorityQueue<Integer> right = new PriorityQueue<>();
        Queue<Integer> mid = new LinkedList<>();

        int n = Integer.parseInt(br.readLine());
        for (int i = 0; i < n; i++) {
            int num = Integer.parseInt(br.readLine());

            if (mid.isEmpty()) {
                mid.add(num);
            } else {
                if (mid.peek() <= num) {
                    right.offer(num);
                } else {
                    left.offer(num);
                }
            }

            if (left.size() > right.size()) {
                right.offer(mid.poll());
                mid.add(left.poll());
            }
            if (right.size() > left.size() + 1) {
                left.offer(mid.poll());
                mid.add(right.poll());
            }

            sb.append(mid.peek()).append('\n');
        }
        System.out.print(sb);
    }
}

File: 백준/Gold/1655. 가운데를 말해요/가운데를 말해요.py
import sys
from heapq import heappush, heappop
input = lambda: sys.stdin.readline().rstrip()
left =[]
mid=[]
right=[]

n = int(input())
for i in range(n):
    num = int(input())
    if not mid:
        mid.append(num)
    else:
        if mid[0]<=num:
            heappush(right,num)
        else:
            heappush(left,num*-1)

    if len(left)>len(right):
        heappush(right,mid.pop())
        mid.append(heappop(left)*-1)
    if len(right)>len(left)+1:
        heappush(left,mid.pop()*-1)
        mid.append(heappop(right))
    print(mid[0])

